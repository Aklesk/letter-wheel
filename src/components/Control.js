import React from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  ListView,
  Vibration,
  View,
  Text,
  TouchableOpacity,
} from 'react-native';
import { inject, observer } from 'mobx-react';
import Icon from 'react-native-vector-icons/Ionicons';

import ControlStyle from '../styles/ControlStyle';

@inject('appStore') @observer
export default class Control extends React.Component {
  static propTypes = {
    appStore: PropTypes.shape({
      dataSource: PropTypes.object.isRequired,
      gameModal: PropTypes.object,
      getScore: PropTypes.func.isRequired,
      letters: PropTypes.string.isRequired,
      scores: PropTypes.array.isRequired,
      selected: PropTypes.object.isRequired,
      statusText: PropTypes.string.isRequired,
      submitWord: PropTypes.func.isRequired,
      timer: PropTypes.number.isRequired,
      tried: PropTypes.object.isRequired,
      words: PropTypes.object.isRequired,
    }).isRequired,
    onCorrect: PropTypes.func,
    onWrong: PropTypes.func,
  }
  static defaultProps = {
    onCorrect() {},
    onWrong() {},
  }
  submitWord = () => {
    const wordValidity = this.props.appStore.submitWord();
    if (wordValidity === true) {
      this.props.onCorrect();
    } else if (wordValidity === false) {
      this.props.onWrong();
    }
  }
  render() {
    const { appStore } = this.props;
    const correct = appStore.tried.filter(tryEntry => (
      appStore.words.indexOf(tryEntry.word) !== -1
    )).length;
    const {
      backspaceTouch,
      backspaceWrapper,
      buttonWrapper,
      columnContainer,
      container,
      entryContainer,
      entryWrapper,
      entryText,
      leftColumn,
      leftColumnHeader,
      leftColumnHeaderText,
      resultContainer,
      resultText,
      rightColumn,
      row,
      progressContainer,
      progressText,
      progressTextSmall,
      timerContainer,
      timerText,
    } = ControlStyle;
    const formattedTimer = appStore.timer / 60 >= 1 ?
      `${Math.floor(appStore.timer / 60)}m ${appStore.timer % 60}s` :
      `${appStore.timer % 60}s`;
    const score = appStore.getScore();
    const feedbackText = (() => {
      const { toNext } = score;
      if (toNext > 0) {
        return `${toNext} word${toNext > 1 ? 's' : ''}\n to next level`;
      }
      return '';
    })();
    return (
      <View style={container}>
        <View style={entryContainer}>
          <View style={entryWrapper}>
            <Text style={entryText}>
              {
                appStore.selected.map(i => appStore.letters[i].toUpperCase()).join('')
              }
            </Text>
          </View>
          <View style={backspaceWrapper}>
            <TouchableOpacity
              onPress={() => { appStore.selected.pop(); }}
              onLongPress={() => {
                Vibration.vibrate(100);
                appStore.selected.clear();
              }}
              style={backspaceTouch}
            >
              <Icon
                name="md-backspace"
                size={35}
              />
            </TouchableOpacity>
          </View>
        </View>
        <View style={columnContainer}>

          <View style={leftColumn}>
            <View style={leftColumnHeader}>
              <Text style={leftColumnHeaderText}>
                {`${correct} / ${appStore.words.length}`}
              </Text>
            </View>
            <ListView
              dataSource={appStore.dataSource}
              enableEmptySections
              renderRow={singleRow => (
                <Text style={ControlStyle[singleRow.style]}>
                  {singleRow.word}
                </Text>
              )}
              style={row}
            />
          </View>

          <View style={rightColumn}>
            <View style={resultContainer}>
              <Text style={resultText}>
                {appStore.statusText}
              </Text>
            </View>
            <View style={buttonWrapper}>
              <Button
                color="#999"
                onPress={this.submitWord}
                title="Submit"
              />
            </View>
            {appStore.timer !== -1 ?
              <View style={timerContainer}>
                <Text style={timerText}>
                  {formattedTimer}
                </Text>
              </View>
              :
              <View />
            }
            <View style={progressContainer}>
              <View>
                <Text style={progressText}>
                  {score.text !== '' ? `${score.text}!` : ' '}
                </Text>
              </View>
              <View>
                <Text style={progressTextSmall}>
                  {feedbackText}
                </Text>
              </View>
            </View>
            <View style={buttonWrapper}>
              <Button
                color="#999"
                onPress={() => {
                  appStore.gameModal.close();
                  appStore.gameModal.open();
                }}
                title="Menu"
              />
            </View>
          </View>

        </View>
      </View>
    );
  }
}
