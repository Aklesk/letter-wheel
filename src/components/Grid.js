import React from 'react';
import PropTypes from 'prop-types';
import {
  ImageBackground,
  Text,
  TouchableOpacity,
  Vibration,
  View,
} from 'react-native';
import { observer } from 'mobx-react';
import * as Animatable from 'react-native-animatable';
import Icon from 'react-native-vector-icons/Ionicons';

import GridStyle from '../styles/GridStyle';

import background from '../images/gridBackground.png';
import select1 from '../images/select1.png';
import select2 from '../images/select2.png';
import select3 from '../images/select3.png';
import select4 from '../images/select4.png';
import select5 from '../images/select5.png';
import select6 from '../images/select6.png';
import select7 from '../images/select7.png';
import select8 from '../images/select8.png';
import select9 from '../images/select9.png';

const customPulseAnimation = {
  0.00: { scale: 1 },
  0.05: { scale: 1 },
  0.30: { scale: 1.05 },
  0.70: { scale: 0.95 },
  0.95: { scale: 1 },
  1.00: { scale: 1 },
};

const customShakeAnimation = {
  0.000: { translateX: 0 },
  0.250: { translateX: -6 },
  0.375: { translateX: 6 },
  0.500: { translateX: -6 },
  0.625: { translateX: 6 },
  0.750: { translateX: -6 },
  1.000: { translateX: 0 },
};

const selectImages = [
  select1,
  select2,
  select3,
  select4,
  select5,
  select6,
  select7,
  select8,
  select9,
];

@observer
export default class Grid extends React.Component {
  static propTypes = {
    animationState: PropTypes.string,
    clearAnimation: PropTypes.func,
    clearSelected: PropTypes.func,
    gridEntries: PropTypes.oneOfType([
      PropTypes.objectOf(PropTypes.shape({
        letter: PropTypes.string,
        selected: PropTypes.bool,
      })),
      PropTypes.arrayOf(PropTypes.shape({
        letter: PropTypes.string,
        selected: PropTypes.bool,
      })),
    ]),
    selectedLetters: PropTypes.string,
    submitWord: PropTypes.func,
    toggleSelected: PropTypes.func,
    triggerAnimation: PropTypes.func,
  }

  static defaultProps = {
    animationState: '',
    clearAnimation() {},
    clearSelected() {},
    gridEntries: Array.from(Array(9)).map(() => ({
      letter: '',
      selected: false,
    })),
    selectedLetters: '',
    submitWord() {},
    toggleSelected() {},
    triggerAnimation() {},
  }

  makeBlock = (i) => {
    const { letter, selected } = this.props.gridEntries[i];
    let style = 'block';
    if (i === '4') {
      style = 'centerBlock';
    }

    return (
      <TouchableOpacity
        activeOpacity={0.8}
        onPress={() => { this.props.toggleSelected(i); }}
        onLongPress={() => {
          Vibration.vibrate(100);
          if (!selected) {
            this.props.toggleSelected(i);
          }
          this.submitWord();
        }}
        style={GridStyle[style]}
        key={`gridItem${i}`}
      >
        <ImageBackground
          source={selected ? selectImages[i] : null}
          style={GridStyle.gridSelect}
        >
          <Text style={GridStyle.letter}>
            {letter.toUpperCase()}
          </Text>
        </ImageBackground>
      </TouchableOpacity>
    );
  }

  clearOne = () => {
    this.props.clearSelected(1);
  }

  clearAll = () => {
    Vibration.vibrate(100);
    this.props.clearSelected(9);
  }

  submitWord = () => {
    const wordValidity = this.props.submitWord();
    if (wordValidity === true) {
      this.props.triggerAnimation('correct');
    } else if (wordValidity === false) {
      this.props.triggerAnimation('incorrect');
    }
  }

  render() {
    const rows = [['0', '1', '2'], ['3', '4', '5'], ['6', '7', '8']].map(block => (
      <View style={GridStyle.row} key={block.join('')}>
        {
          block.map(i => (
            this.makeBlock(i)
          ))
        }
      </View>
    ));
    const {
      backspaceWrapper,
      backspaceTouch,
      container,
      entryContainer,
      entryText,
      entryWrapper,
      gridBackground,
      gridWrapper,
    } = GridStyle;

    const animationType = {};
    switch (this.props.animationState) {
      case 'correct':
        animationType.animation = customPulseAnimation;
        animationType.duration = 1000;
        break;
      case 'incorrect':
        animationType.animation = customShakeAnimation;
        animationType.duration = 500;
        break;
      default:
        animationType.animation = { 0: {}, 1: {} };
        animationType.duration = 0;
        break;
    }

    return (
      <Animatable.View
        animation={animationType.animation}
        duration={animationType.duration}
        iterationCount={1}
        onAnimationEnd={this.props.clearAnimation}
        style={gridWrapper}
      >
        <ImageBackground
          source={background}
          style={gridBackground}
        >
          <View style={container}>
            {rows}
          </View>
          <View style={entryContainer}>
            <View style={entryWrapper}>
              <Text style={entryText}>
                {this.props.selectedLetters}
              </Text>
            </View>
            <View style={backspaceWrapper}>
              <TouchableOpacity
                onPress={this.clearOne}
                onLongPress={this.clearAll}
                style={backspaceTouch}
              >
                <Icon
                  name="md-backspace"
                  size={35}
                />
              </TouchableOpacity>
            </View>
          </View>
        </ImageBackground>
      </Animatable.View>
    );
  }
}
