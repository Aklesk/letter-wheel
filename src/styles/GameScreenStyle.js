import { StyleSheet } from 'react-native';
import { verticalScale } from 'react-native-size-matters';

export default StyleSheet.create({
  portrait: {
    flex: 1,
    flexDirection: 'column',
    width: '100%',
  },
  landscape: {
    flex: 1,
    flexDirection: 'row',
  },
  dataWrapper: {
    flex: 1,
  },
  decorativePaper: {
    position: 'absolute',
    height: verticalScale(375),
    width: verticalScale(400),
    right: verticalScale(-320),
    top: verticalScale(-220),
  },
});
