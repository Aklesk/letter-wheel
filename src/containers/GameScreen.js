import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import { inject, observer } from 'mobx-react';
import nativeTimer from 'react-native-timer';
import * as Animatable from 'react-native-animatable';
import KeyEvent from 'react-native-keyevent';

import WrapperStyle from '../styles/WrapperStyle';
import GameScreenStyle from '../styles/GameScreenStyle';
import Grid from '../components/Grid';
import Control from '../components/Control';

const customPulse = {
  0: {
    scale: 1,
  },
  0.05: {
    scale: 1,
  },
  0.3: {
    scale: 1.05,
  },
  0.7: {
    scale: 0.95,
  },
  0.95: {
    scale: 1,
  },
  1: {
    scale: 1,
  },
};

const customShake = {
  0: {
    translateX: 0,
  },
  0.25: {
    translateX: -6,
  },
  0.375: {
    translateX: 6,
  },
  0.5: {
    translateX: -6,
  },
  0.625: {
    translateX: 6,
  },
  0.75: {
    translateX: -6,
  },
  1: {
    translateX: 0,
  },
};

@inject('appStore') @observer
export default class GameScreen extends React.Component {
  static propTypes = {
    appStore: PropTypes.shape({
      aboutModal: PropTypes.object,
      navigator: PropTypes.object,
      gameModal: PropTypes.object,
      instructionsModal: PropTypes.object,
      orientation: PropTypes.number.isRequired,
      scoreGame: PropTypes.func.isRequired,
      timer: PropTypes.number.isRequired,
    }).isRequired,
  }
  constructor() {
    super();
    this.state = {
      submitState: '',
    };
  }
  componentDidMount = () => {
    const { appStore } = this.props;

    KeyEvent.onKeyUpListener((keyCode) => {
      if (
        (keyCode === 1 || keyCode === 82) &&
        appStore.gameModal.state.isOpen !== true &&
        appStore.aboutModal.state.isOpen !== true &&
        appStore.instructionsModal.state.isOpen !== true
      ) {
        appStore.gameModal.close();
        appStore.gameModal.open();
      }
    });

    nativeTimer.setInterval(
      this,
      'gameTimer',
      () => {
        const { nav } = appStore.navigator.state;

        // Only process the timer if it's active,
        // and the user is focused on the game screen
        if (
          appStore.timer >= 0 &&
          appStore.scored !== true &&
          nav.routes[nav.index].routeName === 'Game' &&
          appStore.gameModal.state.isOpen !== true &&
          appStore.aboutModal.state.isOpen !== true &&
          appStore.instructionsModal.state.isOpen !== true
        ) {
          if (appStore.timer > 0) {
            this.props.appStore.timer = (appStore.timer - 1);
          } else {
            this.props.appStore.timer = -1;
            appStore.scoreGame();
          }
        }
      },
      1000,
    );
  }
  componentWillUnmount = () => {
    nativeTimer.clearInterval(this);
    KeyEvent.removeKeyUpListener();
  }
  onCorrect = () => {
    this.setState({ submitState: 'correct' });
  }
  onWrong = () => {
    this.setState({ submitState: 'incorrect' });
  }
  renderGrid = () => {
    const { gridWrapper } = GameScreenStyle;
    if (this.state.submitState === 'incorrect') {
      return (
        <Animatable.View
          animation={customShake}
          duration={500}
          iterationCount={1}
          onAnimationEnd={() => { this.setState({ submitState: '' }); }}
          style={gridWrapper}
        >
          <Grid
            onCorrect={this.onCorrect}
            onWrong={this.onWrong}
          />
        </Animatable.View>
      );
    } else if (this.state.submitState === 'correct') {
      return (
        <Animatable.View
          animation={customPulse}
          duration={1000}
          iterationCount={1}
          onAnimationEnd={() => { this.setState({ submitState: '' }); }}
          style={gridWrapper}
        >
          <Grid
            onCorrect={this.onCorrect}
            onWrong={this.onWrong}
          />
        </Animatable.View>
      );
    }
    return (
      <View style={gridWrapper}>
        <Grid
          onCorrect={this.onCorrect}
          onWrong={this.onWrong}
        />
      </View>
    );
  }
  render() {
    const { appStore } = this.props;
    const orientation = appStore.orientation === 0 ? 'portrait' : 'landscape';
    const { dataWrapper } = GameScreenStyle;
    const { container } = WrapperStyle;
    return (
      <View style={container}>
        <View style={GameScreenStyle[orientation]}>
          {this.renderGrid()}
          <View style={dataWrapper}>
            <Control
              onCorrect={this.onCorrect}
              onWrong={this.onWrong}
            />
          </View>
        </View>
      </View>
    );
  }
}
