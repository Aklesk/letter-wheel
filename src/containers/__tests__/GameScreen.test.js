import React from 'react';
import { StyleSheet } from 'react-native';
import { mount } from 'enzyme';
import nativeTimer from 'react-native-timer';
import KeyEvent from 'react-native-keyevent';
import { runInAction } from 'mobx';

import AppStore from '../../data/store';
import GameScreen from '../GameScreen';
import Grid from '../../components/Grid';
import Control from '../../components/Control';
import GameScreenStyle from '../../styles/GameScreenStyle';

jest.mock('../../components/Grid', () => {
  require('react'); // eslint-disable-line global-require
  const { View } = require('react-native'); // eslint-disable-line global-require
  return props => (
    <View {...props}>
      {'18E4DC45-B77E-46C2-B7F5-D6AFA39240BB'}
    </View>
  );
});
jest.mock('../../components/Control', () => {
  require('react'); // eslint-disable-line global-require
  const { View } = require('react-native'); // eslint-disable-line global-require
  return props => (
    <View {...props}>
      {'15D067E6-1538-46AD-89A0-8D8729169089'}
    </View>
  );
});
jest.mock('react-native-keyevent', () => ({
  onKeyUpListener: jest.fn(),
  removeKeyUpListener: jest.fn(),
}));

describe('Game Screen component', () => {
  let store;

  beforeEach(() => {
    store = new AppStore();
  });

  afterEach(() => {
    jest.clearAllMocks();
    store = undefined;
  });

  it('matches the snapshot', () => {
    const render = mount(<GameScreen store={store} />);
    expect(render).toMatchSnapshot();
  });

  it('exports the GameScreen component', () => {
    const render = mount(<GameScreen store={store} />);
    expect(render.find(GameScreen)).toHaveLength(1);
  });

  it('adds an event listener to the menu button to open the menu', () => {
    expect(KeyEvent.onKeyUpListener).toHaveBeenCalledTimes(0);
    mount(<GameScreen store={store} />);
    jest.spyOn(store, 'openMenuModal');

    expect(KeyEvent.onKeyUpListener).toHaveBeenCalledTimes(1);
    expect(KeyEvent.onKeyUpListener).toHaveBeenCalledWith(expect.any(Function));

    const listener = KeyEvent.onKeyUpListener.mock.calls[0][0];
    expect(store.openMenuModal).toHaveBeenCalledTimes(0);

    for (let i = 0; i < 255; i += 1) {
      if (i !== 1 && i !== 82) {
        listener(i);
        expect(store.openMenuModal).toHaveBeenCalledTimes(0);
      }
    }

    runInAction(() => { store.isMenuModalOpen = true; });
    listener(1);
    listener(82);
    expect(store.openMenuModal).toHaveBeenCalledTimes(0);

    runInAction(() => {
      store.isMenuModalOpen = false;
      store.isAboutModalOpen = true;
    });
    listener(1);
    listener(82);
    expect(store.openMenuModal).toHaveBeenCalledTimes(0);

    runInAction(() => {
      store.isAboutModalOpen = false;
      store.isInstructionsModalOpen = true;
    });
    listener(1);
    listener(82);
    expect(store.openMenuModal).toHaveBeenCalledTimes(0);

    runInAction(() => { store.isInstructionsModalOpen = false; });

    store.openMenuModal.mockClear();
    listener(1);
    expect(store.openMenuModal).toHaveBeenCalledTimes(1);

    store.openMenuModal.mockClear();
    runInAction(() => { store.isMenuModalOpen = false; });
    listener(82);
    expect(store.openMenuModal).toHaveBeenCalledTimes(1);
  });

  it('handles timers correctly', () => {
    store.nav.get = jest.fn();
    mount(<GameScreen.wrappedComponent store={store} />);
    jest.spyOn(store, 'setTimer');
    jest.spyOn(store, 'scoreGame');

    expect(store.timer).toEqual(-1);
    jest.advanceTimersByTime(1000);
    expect(store.timer).toEqual(-1);
    jest.advanceTimersByTime(10000);
    expect(store.timer).toEqual(-1);

    runInAction(() => { store.timer = 550; });
    store.nav.get.mockReturnValue('Menu');
    jest.advanceTimersByTime(5000);
    expect(store.timer).toEqual(550);

    store.nav.get.mockReturnValue('Game');
    jest.advanceTimersByTime(1000);
    expect(store.timer).toEqual(549);
    jest.advanceTimersByTime(1000);
    expect(store.timer).toEqual(548);
    jest.advanceTimersByTime(500000);
    expect(store.timer).toEqual(48);
    jest.advanceTimersByTime(48000);
    expect(store.timer).toEqual(0);
    expect(store.scoreGame).toHaveBeenCalledTimes(0);
    jest.advanceTimersByTime(1000);
    expect(store.timer).toEqual(-1);
    expect(store.scoreGame).toHaveBeenCalledTimes(1);
  });

  it('unmounts gracefully', () => {
    jest.spyOn(nativeTimer, 'clearInterval');
    const render = mount(<GameScreen.wrappedComponent store={store} />);
    const instance = render.instance();
    expect(nativeTimer.clearInterval).toHaveBeenCalledTimes(0);
    expect(KeyEvent.removeKeyUpListener).toHaveBeenCalledTimes(0);
    render.unmount();
    expect(nativeTimer.clearInterval).toHaveBeenCalledTimes(1);
    expect(nativeTimer.clearInterval).toHaveBeenCalledWith(instance);
    expect(KeyEvent.removeKeyUpListener).toHaveBeenCalledTimes(1);
  });

  it('provides a computed for grid letters', () => {
    const render = mount(<GameScreen.wrappedComponent store={store} />);
    expect(render.instance().gridEntries).toEqual({
      0: { letter: ' ', selected: false },
      1: { letter: ' ', selected: false },
      2: { letter: ' ', selected: false },
      3: { letter: ' ', selected: false },
      4: { letter: ' ', selected: false },
      5: { letter: ' ', selected: false },
      6: { letter: ' ', selected: false },
      7: { letter: ' ', selected: false },
      8: { letter: ' ', selected: false },
    });

    runInAction(() => { store.letters = 'fefbbcbdf'; });
    expect(render.instance().gridEntries).toEqual({
      0: { letter: 'f', selected: false },
      1: { letter: 'e', selected: false },
      2: { letter: 'f', selected: false },
      3: { letter: 'b', selected: false },
      4: { letter: 'b', selected: false },
      5: { letter: 'c', selected: false },
      6: { letter: 'b', selected: false },
      7: { letter: 'd', selected: false },
      8: { letter: 'f', selected: false },
    });

    runInAction(() => { store.letters = 'dfbecfbfb'; });
    runInAction(() => { store.selected.replace(['1', '4', '8']); });
    expect(render.instance().gridEntries).toEqual({
      0: { letter: 'd', selected: false },
      1: { letter: 'f', selected: true },
      2: { letter: 'b', selected: false },
      3: { letter: 'e', selected: false },
      4: { letter: 'c', selected: true },
      5: { letter: 'f', selected: false },
      6: { letter: 'b', selected: false },
      7: { letter: 'f', selected: false },
      8: { letter: 'b', selected: true },
    });
  });

  it('provides animationState with actions to update it', () => {
    const render = mount(<GameScreen.wrappedComponent store={store} />);

    expect(render.instance().animationState).toEqual('');
    expect(render.instance().triggerAnimation).toEqual(expect.any(Function));
    expect(render.instance().clearAnimation).toEqual(expect.any(Function));

    render.instance().triggerAnimation('933ADD04285C');
    expect(render.instance().animationState).toEqual('933ADD04285C');
    render.instance().triggerAnimation('06DAD913982E');
    expect(render.instance().animationState).toEqual('06DAD913982E');

    render.instance().clearAnimation();
    expect(render.instance().animationState).toEqual('');
    render.instance().triggerAnimation('06DAD913982E');
    render.instance().clearAnimation();
    expect(render.instance().animationState).toEqual('');
  });

  it('renders a Grid element', () => {
    runInAction(() => { store.letters = 'dfbecfbfb'; });
    runInAction(() => { store.selected.replace(['1', '4', '8']); });
    const render = mount(<GameScreen.wrappedComponent store={store} />);

    const grid = render.find(Grid).at(0);
    expect(grid.props().animationState).toEqual('');
    expect(grid.props().clearAnimation).toEqual(render.instance().clearAnimation);
    expect(grid.props().clearSelected).toEqual(store.clearSelected);
    expect(grid.props().selectedLetters).toEqual('FCB');
    expect(grid.props().toggleSelected).toEqual(store.toggleSelected);
    expect(grid.props().gridEntries).toEqual(render.instance().gridEntries);
    expect(grid.props().submitWord).toEqual(store.submitWord);
    expect(grid.props().triggerAnimation).toEqual(render.instance().triggerAnimation);
  });

  it('renders a Control element', () => {
    runInAction(() => { store.letters = 'dfbecfbfb'; });
    runInAction(() => { store.selected.replace(['1', '4', '8']); });
    const render = mount(<GameScreen.wrappedComponent store={store} />);
    jest.spyOn(store, 'submitWord');
    jest.spyOn(render.instance(), 'triggerAnimation');

    const control = render.find(Control).at(0);
    expect(control.props().isScored).toEqual(store.scored);
    expect(control.props().onMenu).toEqual(store.openMenuModal);
    expect(control.props().onSubmit).toEqual(expect.any(Function));
    expect(control.props().scoreText).toEqual(store.score.text);
    expect(control.props().statusText).toEqual(store.statusText);
    expect(control.props().statusText).toEqual('');
    expect(control.props().tried).toEqual(store.tried);
    expect(control.props().words).toEqual(store.words);

    store.submitWord.mockReturnValueOnce(true);
    control.props().onSubmit();
    expect(render.instance().triggerAnimation).toHaveBeenCalledTimes(1);
    expect(render.instance().triggerAnimation).toHaveBeenCalledWith('correct');

    jest.clearAllMocks();
    store.submitWord.mockReturnValueOnce(false);
    control.props().onSubmit();
    expect(render.instance().triggerAnimation).toHaveBeenCalledTimes(1);
    expect(render.instance().triggerAnimation).toHaveBeenCalledWith('incorrect');
  });

  it('responds to screen orientation', () => {
    const render = mount(<GameScreen.wrappedComponent store={store} />);
    let orientationWrapper;

    orientationWrapper = render.find({ style: GameScreenStyle.portrait }).at(0);
    expect(orientationWrapper).toHaveLength(1);
    expect(orientationWrapper.props().style).toEqual(StyleSheet.flatten(GameScreenStyle).portrait);

    runInAction(() => { store.orientation = 1; });
    render.update();
    orientationWrapper = render.find({ style: GameScreenStyle.landscape }).at(0);
    expect(orientationWrapper).toHaveLength(1);
    expect(orientationWrapper.props().style).toEqual(StyleSheet.flatten(GameScreenStyle).landscape);
  });
});
