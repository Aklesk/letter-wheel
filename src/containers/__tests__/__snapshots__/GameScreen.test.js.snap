// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Screen component matches the snapshot 1`] = `
<inject-GameScreen-with-store
  store={
    store {
      "closeAllModals": [Function],
      "db": Object {
        "transaction": [MockFunction],
      },
      "getPermutatedWords": [Function],
      "isAboutModalOpen": false,
      "isInstructionsModalOpen": false,
      "isMenuModalOpen": false,
      "letters": "         ",
      "loading": false,
      "nav": Object {
        "get": [Function],
        "goto": [Function],
        "resetto": [Function],
      },
      "navigator": Object {},
      "newGameOptions": Object {
        "timed": false,
        "wordRange": Object {
          "max": 49,
          "min": 10,
        },
      },
      "orientation": 0,
      "scored": false,
      "scores": Array [
        Object {
          "percent": 100,
        },
        Object {
          "percent": 85,
        },
        Object {
          "percent": 65,
        },
        Object {
          "percent": 40,
        },
        Object {
          "percent": 0,
        },
      ],
      "selected": Array [],
      "shuffle": [Function],
      "statusText": "",
      "timer": -1,
      "tried": Object {},
      "width": 320,
      "words": Object {},
    }
  }
>
  <GameScreen
    store={
      store {
        "closeAllModals": [Function],
        "db": Object {
          "transaction": [MockFunction],
        },
        "getPermutatedWords": [Function],
        "isAboutModalOpen": false,
        "isInstructionsModalOpen": false,
        "isMenuModalOpen": false,
        "letters": "         ",
        "loading": false,
        "nav": Object {
          "get": [Function],
          "goto": [Function],
          "resetto": [Function],
        },
        "navigator": Object {},
        "newGameOptions": Object {
          "timed": false,
          "wordRange": Object {
            "max": 49,
            "min": 10,
          },
        },
        "orientation": 0,
        "scored": false,
        "scores": Array [
          Object {
            "percent": 100,
          },
          Object {
            "percent": 85,
          },
          Object {
            "percent": 65,
          },
          Object {
            "percent": 40,
          },
          Object {
            "percent": 0,
          },
        ],
        "selected": Array [],
        "shuffle": [Function],
        "statusText": "",
        "timer": -1,
        "tried": Object {},
        "width": 320,
        "words": Object {},
      }
    }
  >
    <ImageBackground
      source={
        Object {
          "testUri": "src/images/woodBackground.jpg",
        }
      }
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "flex": 1,
          "height": null,
          "justifyContent": "center",
          "width": null,
        }
      }
    >
      <react-native-mock>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "width": "100%",
            }
          }
        >
          <react-native-mock>
            <Image
              source={
                Object {
                  "testUri": "src/images/tiltedPanel.png",
                }
              }
              style={
                Object {
                  "height": 423.52941176470586,
                  "position": "absolute",
                  "right": -361.4117647058823,
                  "top": -248.47058823529412,
                  "width": 451.7647058823529,
                }
              }
            >
              <react-native-mock />
            </Image>
            <Component
              animationState=""
              clearAnimation={[Function]}
              clearSelected={[Function]}
              gridEntries={
                Object {
                  "0": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "1": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "2": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "3": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "4": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "5": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "6": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "7": Object {
                    "letter": " ",
                    "selected": false,
                  },
                  "8": Object {
                    "letter": " ",
                    "selected": false,
                  },
                }
              }
              selectedLetters=""
              submitWord={[Function]}
              toggleSelected={[Function]}
              triggerAnimation={[Function]}
            >
              <View
                animationState=""
                clearAnimation={[Function]}
                clearSelected={[Function]}
                gridEntries={
                  Object {
                    "0": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "1": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "2": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "3": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "4": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "5": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "6": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "7": Object {
                      "letter": " ",
                      "selected": false,
                    },
                    "8": Object {
                      "letter": " ",
                      "selected": false,
                    },
                  }
                }
                selectedLetters=""
                submitWord={[Function]}
                toggleSelected={[Function]}
                triggerAnimation={[Function]}
              >
                <react-native-mock>
                  18E4DC45-B77E-46C2-B7F5-D6AFA39240BB
                </react-native-mock>
              </View>
            </Component>
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <react-native-mock>
                <Component
                  isScored={false}
                  onMenu={[Function]}
                  onSubmit={[Function]}
                  scoreRank={4}
                  statusText=""
                  timerString=""
                  tried={Object {}}
                  words={Object {}}
                  wordsToNextLevel={0}
                >
                  <View
                    isScored={false}
                    onMenu={[Function]}
                    onSubmit={[Function]}
                    scoreRank={4}
                    statusText=""
                    timerString=""
                    tried={Object {}}
                    words={Object {}}
                    wordsToNextLevel={0}
                  >
                    <react-native-mock>
                      15D067E6-1538-46AD-89A0-8D8729169089
                    </react-native-mock>
                  </View>
                </Component>
              </react-native-mock>
            </View>
          </react-native-mock>
        </View>
      </react-native-mock>
    </ImageBackground>
  </GameScreen>
</inject-GameScreen-with-store>
`;
